docs/core/testing/mstest-analyzers/mstest0040.md

```diff
@@ -0,0 +1,41 @@
+---
+title: "MSTEST0040: Do not assert inside 'async void' contexts"
+description: "Learn about code analysis rule MSTEST0040: Do not assert inside 'async void' methods, local functions, or lambdas because they may not fail the test"
+ms.date: 01/17/2025
+f1_keywords:
+- MSTEST0040
+- AvoidUsingAssertsInAsyncVoidContextAnalyzer
+helpviewer_keywords:
+- AvoidUsingAssertsInAsyncVoidContextAnalyzer
+- MSTEST0040
+author: Youssef1313
+ms.author: ygerges
+---
+# MSTEST0040: Do not assert inside 'async void' contexts
+
+| Property                            | Value                                                                  |
+|-------------------------------------|------------------------------------------------------------------------|
+| **Rule ID**                         | MSTEST0040                                                             |
+| **Title**                           | Do not assert inside 'async void' contexts                             |
+| **Category**                        | Usage                                                                  |
+| **Fix is breaking or non-breaking** | Non-breaking                                                           |
+| **Enabled by default**              | Yes                                                                    |
+| **Default severity**                | Warning                                                                |
+| **Introduced in version**           | 3.8.0                                                                  |
+| **Is there a code fix**             | No                                                                     |
+
+## Cause
+
+The use of any assertion method in an `async void` method, local function, or lambda.
+
+## Rule description
+
+Exceptions that are thrown in an `async void` context are unobserved. So, a failing assertion in an `async void` method may go unnoticed. When using VSTest, such exceptions are very likely to be unnoticed. When using Microsoft.Testing.Platform, such exception *may* crash the process
```

Comment:
@stephentoub I'll appreciate if you can review the wording for accuracy here please.

To share context: in the case of Microsoft.Testing.Platform, we do subscribe to both `TaskScheduler.UnobservedTaskException` and `AppDomain.CurrentDomain.UnhandledException` and call `FailFast`. In VSTest, we don't subscribe.

Reply:
How do you prevent the crashing in VSTest? Is there a SynchronizationContext that's eating the exception? Are unhandled exceptions on ThreadPool threads configured to not crash?

Exceptions in an async void method are just queued to be thrown on the current SynchronizationContext if there is one, or if there isn't on the ThreadPool.

Reply:
Hmm.. VSTest doesn't have synchronization context. Not sure about the second question of configuring ThreadPool threads to not crash. @Evangelink @nohwnd Do you know?

@Evangelink Could it be that the test project where the original issue is noticed is setting their own synchronization context? 👀

Reply:
I'll look at VSTest codebase when I'm back at laptop.

Reply:
I tried a quick sample in a VSTest project. The test passes when `async void` which is kinda expected even for MTP, but I can see the UnhandledException being printed with VSTest.

Maybe it's just about that FailFast in MTP which will simply cause non-zero exit code and things will be failing properly in CI contexts etc?

Reply:
@stephentoub could this `<legacyUnhandledExceptionPolicy enabled="1" />` have an impact? https://github.com/microsoft/vstest/blob/71a1cc4b3896bff845a28658f71d6886b8d2ea44/src/testhost.x86/app.config#L7

Reply:
Original issue was on netfx so this is pretty sure it's linked.

Reply:
Yes, that's what I meant by "Are unhandled exceptions on ThreadPool threads configured to not crash?" That setting reverts to the .NET Framework 1.x behavior  circa 20 years ago where the ThreadPool eats unhandled exceptions rather than allowing them to crash the process.

Reply:
Is the following wording more accurate then?

```
Exceptions that are thrown in an `async void` context are unhandled. A failing assertion in an `async void` method will be swallowed and will not crash the process when using VSTest under .NET Framework, and may be unnoticed when using Microsoft.Testing.Platform or VSTest under modern .NET .
```

Reply:
Seems reasonable I'd guess. I'd find the behavior more intuitive if it behaved like xunit does: xunit registers its own SynchronizationContext in order to make `async void` work, both in terms of knowing when the test completes and knowing when an exception occurs. `async void` still isn't recommended, but having one doesn't put the whole process in jeopardy.

Reply:
@stephentoub Interesting. Let me open an issue for that on our side.